---
# nolint: line_length_linter.
title: "Atlas Output of Mouse Metagenome."
# nolint: line_length_linter.
output:
  learnr::tutorial:
    progressive: true
    allow_skip: true
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
```
## QC report


The `reports/QC_report.html` gives a graphical report on the most important numbers.

The data set used for the Tutorial is a very small one, here you can see the [QC report](https://metagenome-atlas.readthedocs.io/en/latest/_static/QC_report.html) of a bigger run.

<!-- TODO: change reports here -->


```{r qcreport2, echo=FALSE}
message <- "Sample SRS1992976 has very few final reads.
Even more striking is that it started from a high number of initial reads."
question(
    "In the bigger run, are all samples ok? ",
    answer("yes", message = message),
    answer("no", message = message, correct = TRUE)
)
```


### Low quality samples

```{r lowq-sample, echo=FALSE}
message <- "During the contamination and host removal.
Given that the sample comes from a diarrhea patients does this make sense?"
question(
    "During which stage did the sample SRS1992976 lost most of the reads?",
    answer("raw", message = message),
    answer("deduplicated", message = message),
    answer("filtered", message = message),
    answer("qc", message = message, correct = TRUE)
)
```



```{r lowq-sample2, echo=FALSE}
message <- "It depends on your project.
To remove it you can simply drop the line in the sample.tsv.
Note: In the current version of atlas low abundant samples that might give no bin results and stop the pipeline.
In the future it should be easier to exclude samples from binning.
But include it in the quantification."
question(
    "What would you do with the sample?",
    answer("Keep it, but check it at the end?",
        message = message, correct = TRUE
    ),
    answer("Drop it, it doesn't make sense to assemble it?",
        message = message, correct = TRUE
    ),
    type = "single"
)
```


### Insert size

The qc report, shows you also the insert size and the read lengths for each sample.
The insert size is the length of the DNA fragment sequenced usually from both ends.

![](images/readlength.jpg)



```{r insert-size, echo=FALSE}
question(
    "On average, do the read pairs overlap?",
    answer("no"),
    answer("yes", correct = TRUE),
    answer("sometimes", correct = TRUE),
    message = "On average, the insert size is ~225 and the read length ~120.
2x 120 >= 225. So the reads cover the insert size. ",
    type = "single"
)
```

```{r insert-size2, echo=FALSE}
message <- "Assembly works fine or even betterwhen there is a gap between the reads."
question(
    "Is this a problem?",
    answer("no", correct = TRUE, message = message),
    answer("yes", message = message)
)
```

_Note: If you have log reads, you can manually add longer k-mer size for the assembly.
See the documentation of the specific assemblers. _
[_Further reading material_](https://www.badgrammargoodsyntax.com/compbio/2018/3/5/compbio-020-reads-fragments-and-inserts-what-you-need-to-know-for-understanding-your-sequencing-data-1)





## Assembly report

Let's look at the [assembly report](https://metagenome-atlas.readthedocs.io/en/latest/_static/assembly_report.html) of the same (larger) dataset. 



Are you happy with the fraction of assembled reads?



```{r assreport, echo=FALSE}
message <- "Are you suprised that it's not the sample with the least number or reads?"

wrong_message <- paste("The sample with the least Total BP in figure 1.", message)

question(
    "Which sample has the smallest assembly?",
    answer("SRS1992976, the sample with the leas reads.", message = wrong_message),
    answer("SRS1993058", message = wrong_message),
    answer("SRS1992994", correct = TRUE, message = message),
    answer("SRS1992993", message = wrong_message)
)
```


### N50

N50 is a measure for continuity.

The N50-number characterizes how many contigs do you need to constitute 50% of your assembly (starting from the longest). Hence the lower the better. 

The N50-length is the length of the contig at the 50%- limit.
Meaning 50% of your assembly is that long or longer. Hence the larger the better.  

There is some confusion in the field but N50 usually refers to the N50-length.




```{r n50, echo=FALSE}
message <- "Having a high N50-length is a good sign. As it measures the continuity.
However the sample in question has one of the smallest assemblies.
So other samples that have much larger assemblies even with less continuity contain more contigs for binning. "

question(
    "The sample SRS1992993 has by far the highest N50-length. Is this good or bad?",
    answer("It is good, the sample with the leas reads.", message = message, correct = T),
    answer("It is bad", message = message)
)
```


### N90

N90 is the same as N50 but for 90%. 

```{r n90, echo=FALSE}
message <- "All samples have view large and many small contigs, but for this sample it seems to be extreme."


question(
    "Surprisingly, the sample with the highest N50 has very low N90 length. What does this mean?",
    answer("It has a very small assembly.", message = message),
    answer("It has a view long contigs and many small ones.", correct = T, message = message),
    answer("It has very high continuity", message = message)
)
```





## Quality metrics

### What is a good MAG?

A MAG should have at least 50% completeness and less than 10% contamination otherwise it's considered low quality.

Unofficially, we define a high quality MAG as a genome with <5% Contamination and >90% Completeness.
However the official definition contains more.

➤ Use ChatGPT, Bing or Google to find the official definition of a high quality MAG or at least the paper that should contain the definition.


➤ Copy the url to the pdf of the paper

➤ Go to https://www.chatpdf.com/

➤ Ask the AI the same question. 

### Official definitions

Obviously, in publication we stick to the official definition and we use other words like 'near complete' 
or 'Good quality' genomes for the MAGS that have >90% completeness and less 5% contamination. 


```{r neglect, echo=FALSE}
question("Can you imagine why whe neglect the criteria for the presence of t-RNA/ r-NA genes?",
    answer("Yes", correct = TRUE),
    answer("No"),
    message = "They are difficult to assemble."
)
```



### Quality score

In atlas I define a quality score:

$$ \text{Quality score} = \text{Completeness} - 5 \times \text{Contamination} $$

This allows me to have a linear scale to sort the MAGs by quality.

Hence Quality score > 50 is a bit more stringent than 50% completeness and <10% contamination.
Hence Quality score > 90 is a bit more stringent than the good quality genomes.



### Additional metrics

```{r quality, echo=FALSE}
question(
    "What other metric could be of importance to estimate the quality of a MAG?",
    answer("Continuity", correct = TRUE),
    answer("Number of ambiguous bases", correct = TRUE),
    answer("GC content", correct = TRUE),
    answer("Number of contigs", correct = F),
    answer("Number of genes", correct = TRUE),
    answer("Genome size", correct = TRUE)
)
```


## Compare binners
Open the two bin reports.

[DASTool](https://metagenome-atlas.readthedocs.io/en/latest/_static/bin_report_DASTool.html) and
[Vamb](https://metagenome-atlas.readthedocs.io/en/latest/_static/bin_report_vamb.html) in two tabs.






Which binner is better and why?






<!--

https://igv.org/app/

cd WD/sample2/

samtools index sequence_alignment/sample2.bam
samtools faidx sample2_contigs.fasta


import genome,
import track

-->


