---
title: "Analyze the output of metagenome-atlas"
output: html_document
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, 
                      message = FALSE,
                      warning = FALSE, 
                      cache = FALSE, include = TRUE)

```

Load libraries

```{r, libraries,include=FALSE}
library(plotly)
library(heatmaply)
library(dplyr) # dpyr masks select from plotly
library(readr)
library(stringr)
library(tidyr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(pheatmap)
library(microbiome)
library(ape)
library(vegan)
library(useful)
library(kableExtra)
library(ggtree)
```

Atlas output files are stored in the `Example` folder.

```{r, file_paths}
atlas_wd_folder = "../Example/"

taxonomy_file = paste0(atlas_wd_folder,"genomes/taxonomy/gtdb_taxonomy.tsv")
tree_file = paste0(atlas_wd_folder,"genomes/tree/gtdbtk.bac120.nwk")
quality_file= paste0(atlas_wd_folder,"genomes/checkm/completeness.tsv")
counts_file= paste0(atlas_wd_folder,"genomes/counts/raw_counts_genomes.tsv")
abundance_file = paste0(atlas_wd_folder,"genomes/counts/median_coverage_genomes.tsv")
readstats_file= paste0(atlas_wd_folder,"stats/read_counts.tsv")
keggmodules_file = paste0(atlas_wd_folder,"genomes/annotations/dram/kegg_modules.tsv")
```

# Taxonomy

```{r, read_tax}
Tax <- read_tsv(taxonomy_file,show_col_types = FALSE)
kable(Tax)
```

```{r, lab_tax}
# create a short label for each species
Tax <- Tax %>%
  mutate(Label = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Label = ifelse(is.na(Label), paste0(genus, " ", user_genome), Label))
```
## Draw tree

```{r, read_newick}
T <- ggtree::read.tree(tree_file)
T <- full_join(T, rename(Tax,label=user_genome), by = "label")
```



```{r, highlight_tree, fig.height = 10}
library(phytools)

tree_plot <- ggtree(T, layout = 'circular') +
  geom_tiplab(aes(label=Label))

for (phy in unique(Tax$phylum))
{
  phy_tax= Tax %>% filter( phylum == phy)
  if (dim(phy_tax)[1] >1)
  {
    phy_mrca <- findMRCA(T@phylo, phy_tax$user_genome)
    tree_plot <- tree_plot + geom_cladelabel(phy_mrca,
                                              label = phy,
                                              angle=0,
                                              align=TRUE,
                                              fontsize=20,
                                             offset = 0.5,
                                              offset.text=.1,
                                              barsize=5,
                                              )
  }
}

tree_plot
  #geom_highlight(Bacteroidota, fill = "steelblue", extend = 3)
```



# Genome quality

```{r, read_genome_quality}
genome_quality <- read_tsv(quality_file,show_col_types = FALSE) %>%
  mutate(Quality_Score = Completeness - (5*Contamination)) %>%
  left_join(Tax, by = c("Bin Id" = "user_genome")) %>%
  mutate(Name = Label) %>%
  dplyr::select(-Label)
```

```{r, plot_contam_vs_quality}
plt <- ggplot(genome_quality, aes(x = Contamination, y = Completeness, color = phylum, 
                                  text =      paste("Bin Id:", `Bin Id`, "\nName",Name,"\nQuality score:", Quality_Score))) +
  geom_point() + xlab('Contamination [%]') + ylab('Completeness [%]') +
  theme_minimal()
ggplotly(plt,tooltip = c('text' ))


```

# Abundance

## Calculate mapping rate

```{r, mapping_rate, fig.width=1, fig.height=2}
Counts <- read_tsv(counts_file, show_col_types = FALSE)
read_stats <-read_tsv(readstats_file, show_col_types = FALSE) %>% filter(Step == "QC" )

d <- tibble( Sample= colnames(Counts)[-1],
            Mapped_Reads = colSums(Counts[,-1]),
             x=0
             ) %>%
    left_join(read_stats, on = "Sample") %>%
  mutate(Mapping_rate = Mapped_Reads / (Reads_pe*2+Reads_se) *100)

plt <- ggplot(d, aes(y = Mapping_rate, x=x,text=paste("Sample:",Sample))) +
  ylim(c(0, 100))+
  xlim(c(-0.1,0.1)) +
   geom_beeswarm() + 
  theme_minimal()

plt

#ggplotly(plt,tooltip = c('text','y' ))

```




## Relative abundance

For the relative abundance, we take the coverage over the genome, not the raw counts. This implicitly normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.

```{r}
D <- read_tsv(abundance_file, show_col_types = FALSE) %>%
  column_to_rownames(var = "...1") %>%
  as.data.frame()
kable(topleft(D, c= 10))
```

```{r, relative_abund}
# calculate relative abundance
rel_ab <- sweep(D, 1, rowSums(D),`/`)

# get most abundant genomes
counts_per_genome <- data.frame(sums = colSums(rel_ab)) %>%
  rownames_to_column(var = "Sample") %>%
  left_join(Tax, by = c("Sample" = "user_genome")) %>%
  arrange(desc(sums))


ggplot(counts_per_genome %>%
         top_n(sums, n = 10), aes(x = reorder(Label, -sums), y = sums)) +
  geom_col() +
  labs(x = "", y = "Abundance [rel_ab]", title = "Most abundant genomes") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90))
```

### Typical bar chart

```{r, barchart}
level <- 'family'

grouping_columns= c("Sample",level)
grouping_columns <- lapply(grouping_columns, as.symbol)

grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "user_genome") %>%
  pivot_longer(cols = -user_genome, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "user_genome") %>%
  group_by_(.dots = grouping_columns) %>%
  summarise(summarized_rel_ab = sum(rel_ab))

ggplot(grouped_data, aes_string(x = "Sample", y = "summarized_rel_ab", fill = level)) +
  geom_col() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) +
  #scale_fill_manual(values=rep(brewer.pal(5,"Paired"), times=5))
  scale_fill_manual(values = c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788"))

```

# Functional annotations

The relative abundance of functional annotations per sample

The abundance is calculated as the sum of the relative abundance of all bacteria containing a function.

## Kegg modules produced by Dram

```{r, load_modules}

kegg_modules= read_tsv(keggmodules_file,col_select = -1)
module_names= kegg_modules %>% dplyr::select(c('module','module_name')) %>% distinct()


kable(head(kegg_modules))

```

```{r, module_presence}

step_coverage_threshold= 0.8

module_step_coverage_matrix = pivot_wider(kegg_modules,  
                                   id_cols = genome,
                                   names_from = module, 
                                   values_from = step_coverage
                                   ) %>%
                              column_to_rownames("genome") %>% as.matrix() 

module_step_coverage_matrix = module_step_coverage_matrix[, colSums(module_step_coverage_matrix) > 0]


hatmap<- pheatmap(module_step_coverage_matrix, show_colnames = F)

module_presence_matrix = 1 * (module_step_coverage_matrix>step_coverage_threshold)

module_presence_matrix = module_presence_matrix[, colSums(module_presence_matrix) > 0]

hatmap
```

```{r,heatmap_interactive2}



d_tax= column_to_rownames(Tax,"user_genome")



annotations <- module_step_coverage_matrix
for (genome in rownames(annotations))
{
  for (module in colnames(annotations))
  {
    annotations[genome,module] = paste0("Name: ", d_tax[genome,"Label"],
                       "\nPhylum: ",d_tax[genome,"phylum"],
                       "\nPathway: ", module_names[module,"module_name"]
                       )
  }
}




heatmaply(module_step_coverage_matrix,
          custom_hovertext = annotations,
          showticklabels = c(FALSE,FALSE)
          )


```


```{r, module_abundance}
# Sum of rel_ab for all species where a module is presence is equel to the matrix multiplication
stopifnot(dim(rel_ab)[2] == dim(module_presence_matrix)[1]  )
module_rel_ab <- as.matrix(rel_ab) %*% module_presence_matrix

pheatmap(module_rel_ab, show_colnames = F)

```



