---
title: "Differential Abundance Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE,
                      warning = FALSE, cache = FALSE, include = TRUE)
knitr::opts_knit$set(root.dir = "..") # set root dir one level above where Rmd is saved
```

```{r libraries}
library(dplyr)
library(readr)
library(tibble)
library(tidyr)
library(purrr)
library(broom)
library(pheatmap)
library(plotly)
library(microbiome)
library(knitr)
library(ALDEX2)
```

# How does high fat diet impact the microbiome of mice?

![Picture of obese mice](https://upload.wikimedia.org/wikipedia/commons/0/0b/Fatmouse.jpg)

```{r load_metadata}
# load metadata
metadata <- read_tsv('Example/metadata.txt')
kable(metadata)
```

We confirm that the mice on high fat diet really put more weight on.

```{r plot_metadata}
ggplot(metadata, aes(x = Diet, y = Body_weight)) +
  geom_point() +
  theme_minimal()
```

```{r labels}
# create a short label for each genome
Tax <- read_tsv('Example/Results/taxonomy.tsv') %>%
  mutate(Labels = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Labels = ifelse(is.na(Labels), paste0(genus, " ", user_genome), Labels))
```

## Relative abundance

For the relative abundance we take the coverage over the genome not the raw counts. This inmplicit normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.

```{r}
D <- read_tsv("Example/Results/counts/median_coverage_genomes.tsv") %>%
  column_to_rownames(var = "X1")
```


```{r}
# calculate relative abundance
rel_ab <- D/rowSums(D)
```

### Bar chart wich group labels

```{r}
level <- 'family'

grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "user_genome") %>%
  pivot_longer(cols = -user_genome, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "user_genome") %>%
  left_join(metadata, by = c("Sample" = "X1")) %>%
  group_by(Sample, family, Diet) %>%
  summarise(summarized_rel_ab = sum(rel_ab))

ggplot(grouped_data, aes(x = Sample, y = summarized_rel_ab, fill = family)) +
  geom_col() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.title.y = element_blank()) +
  facet_wrap(~Diet, scales = "free_x")
```

We see that the high fat diet induces a big change from the *Muribaculaceae* family (Bacteroidetes) to the *Lachnospiraceae* family.

## Compositional data analysis


In order to analyze the microbiome at the species or genome level we use compositional data analyis (CoDa), see more on [Wikipedia](https://en.wikipedia.org/wiki/Compositional_data) and this article:

>Gloor, Gregory B., Jean M. Macklaim, Vera Pawlowsky-Glahn, and Juan J. Egozcue. 2017. “Microbiome Datasets Are  Compositional: And This Is Not Optional.” Frontiers in Microbiology 8 (November). Frontiers: 2224.
    doi: [10.3389/fmicb.2017.02224](https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224).


For this we load the rawcounts and use centric log ratios (clr) after imputing values for the zeros.




```{r}
#load raw counts

Counts <- read_tsv('Example/Results/counts/raw_counts_genomes.tsv') %>%
  column_to_rownames(var = "Sample") %>%
  t()

# transforme counts with centrig log ratio
data <- transform(Counts, transform = "clr")
```

### PCA (PCoA) of the robust aitchison distance

```{r pca}
transformed_data <- prcomp(data)
```

```{r plot_pca}
pca_data <- transformed_data$x %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(metadata, by = c("Sample" = "X1"))

ggplot(pca_data, aes(x = PC1, y = PC2, color = Diet)) +
  geom_point() +
  theme_minimal() +
  scale_color_manual(values = c(chow = "#00BFC4", HF = "#F8766D"))
```

### Differencial abundance analyis

As the counts are normalized in centered log ratio the log FC becomes the difference.

We use the welch test to assess differential abundance in the two groups. This is a simple version of aldex2. See Gloor et al for more information.

```{aldex2}

x <- aldex.clr(reads= t(Counts), conds= metadata$Diet,mc.samples=16, denom="all", verbose=F)

d.eff <- aldex.effect(x)
d.tt <- aldex.ttest(x)

aldex_stats <- data.frame(d.eff,d.tt)


```


```{r}
# mean abundance per group
Stats <- data %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  pivot_longer(cols = -Sample, names_to = "Id", values_to = "clr") %>%
  left_join(metadata, by = c("Sample" = "X1")) %>%
  group_by(Diet, Id) %>%
  summarize(mean_clr = mean(clr)) %>%
  pivot_wider(id_cols = Id, names_from = Diet, values_from = mean_clr)

# calculate logFC
Stats <- Stats %>%
  mutate(logFC = HF - chow)

# format data for t test
data_ttest <- data %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  pivot_longer(cols = -Sample, names_to = "Id", values_to = "clr") %>%
  left_join(metadata, by = c("Sample" = "X1"))

# run t test
data_ttest <- data_ttest %>%
  group_by(Id, Diet) %>%
  nest() %>%
  spread(key = Diet, value = data) %>%
  mutate(
    t_test = map2(HF, chow, ~{t.test(.x$clr, .y$clr) %>% tidy()}),
    HF = map(HF, nrow),
    chow = map(chow, nrow)
  ) %>%
  unnest() %>%
  select(Id, p.value) %>%
  mutate(Pvalue = p.value) %>%
  mutate(logP = -log10(Pvalue))

Stats <- Stats %>%
  left_join(data_ttest, by = "Id") %>%
  left_join(Tax, by = c("Id" = "user_genome"))
```

#### heatmap of significant Genomes

Correcting form multiple testing would be even better

```{r heatmap, fig.height=8}
# filter to MAG abundances that were significantly different
sig_data <- data %>%
  as.data.frame() %>%
  select(Stats[Stats$Pvalue < 0.01, ]$Id) %>%
  t()

# make a dataframe to use to annotate the heatmap
annot_df <- data.frame(Sample = colnames(sig_data)) %>%
  left_join(metadata, by = c("Sample" = "X1")) %>%
  column_to_rownames(var = "Sample") %>%
  select(Diet)

# sort labels by sig_data order
heatmap_labels <- Tax %>%
  filter(user_genome %in% rownames(sig_data))
heatmap_labels <- heatmap_labels[order(match(heatmap_labels$user_genome, rownames(sig_data))), ]

pheatmap(sig_data, cluster_rows = T, cluster_cols = T, annotation_col = annot_df,
         labels_row = heatmap_labels$Labels)
```

#### Volcano plot

```{r}
## non interactive plot
# ggplot(Stats, aes(x = logFC, y = logP, alpha = logP)) +
#  geom_point(color = "#67000d") +
#  theme_minimal()

plt <- ggplot(Stats, aes(x = logFC, y = logP, alpha = logP,
                         label = Labels, label2 = HF, label3 = chow)) +
  geom_point(color = "#67000d") +
  theme_minimal()

ggplotly(plt, tooltip = c("label", "label2", "label3"))
```

The uncultured species with the name 'UBA7173 sp001689485' is highly significantly increased in chow mice vs HF mice. It belongs to the Muribaculaceae family.

```{r}
ggplot(data %>%
         as.data.frame %>%
         rownames_to_column(var = "Sample") %>%
         left_join(metadata, by = c("Sample" = "X1")),
         aes(y = MAG10, x = Diet, fill = Diet)) +
  geom_boxplot() +
  theme_minimal() +
  scale_fill_manual(values = c(chow = "#00BFC4", HF = "#F8766D"))

kable(Tax %>%
  filter(user_genome == 'MAG10'))
```
